### ISSUE TITLE
Feature request: concurrent downloads of multiple video chunks

### ISSUE THREAD

**Issue by zmwangx** [soimort/you-get#510]
Most Chinese video sites deliver HD videos in independent Flash video chunks, so we can download multiple video chunks concurrently to maximize bandwidth utilization. For instance, in the case of youku.com, I've found (under my normal network conditions) that opening one connection usually leads to a download speed of 20–50 Mbps, while opening ten or more connections at once leads to total download speeds as high as 200–500 Mbps, which is insanely great ®.

I propose that we add a `-j [jobs], --jobs[=jobs]` option, where `jobs` is the number of jobs (i.e., downloads) to run in parallel, and when `-j` is given without argument, do not limit the number of jobs (analogous to the `-j` option of GNU Make). In that case, progress information is harder to print, though; maybe we can get away with just printing the number of finished jobs.

**Comment by linhua55** [comment]
Did you test whether the "Multi-thread downloading" achieve full download speeds?
Refer to #158 #241

**Comment by zmwangx** [comment]
@linhua55 Uh, didn't notice #158. This is basically a duplicate then. Thanks for the reference.

> Did you test whether the "Multi-thread downloading" achieve full download speeds?

What do you mean by "multi-thread downloading"? I don't see this feature in the documentation, or in the source code (upon brief inspection). To achieve the speed I described in my first post, I usually just run multiple background `wget` or `curl` processes in parallel. To do the same in `you-get`, which currently use `urllib` for downloads, I suppose we can simply open a `multiprocessing.Pool`. Or maybe use `threading` instead, but I usually prefer `multiprocessing`. Haven't tested, but I assume this should yield the same level of speed as multiple parallel `wget` or `curl` processes. I'll need to understand the core of the code base to write a multiprocessing patch and do field tests, which unfortunately won't happen any time soon due to my tight schedule.

While we are at it, I don't think supporting `wget` or `aria2` is a good idea. IMHO, Python project should stay as pythonic as possible until it's necessary to use other tools (e.g., FFmpeg). Those who would like to use other download tools could simply write a wrapper, or use the `--url` option to extract bare URIs.

**Comment by linhua55** [comment]
Multi-thread downloading means downloading a file in multi-segments.
For example
>aria2c -x 4 http://k.youku.com/player/getFlvPath/sid/9427620880948122bef7f_00/st/flv/fileid/030002100051D4531BE14405CF07DD46D1E90B-DAB6-BA76-3252-99CDEB847AFF\?K\=d3ee7b68aec9ed26282a5caf\&ctype\=12\&ev\=1\&oip\=1931322792\&token\=9392\&ep\=eyaWHU%2BPVsYI5SrbgD8bZX63InBZXP4J9h%2BFg9NiALshSZy7nD3QteixSvtCE4pvciVyZO%2F339GSY0BmYYQy224QrT%2BvP%2Frj%2FPXh5aRat%2BR1FBEydbyhwlSXQDnz

Which downloading a file in 4 segments.And they are concurrent.
But I ignored the size of chunked file is always small.Thus your idea is better than multi-thread downloading.

**Comment by zmwangx** [comment]
@linhua55 Uh, I see — previously I misinterpreted your point. Yes, assuming the server supports range retrieval requests, we can certainly subdivide each file into segments and retrieve the segments in parallel. However, as you have noted, for sites like Youku, the entire video is already divided into 5—50 MB (from my experience) FLV chunks on the server side, which are already pretty small in size, so the overhead of segmented downloads might defeat the speed boost (or make the speed boost less impressive).

Instead, we can simply download multiple FLV files in parallel, which imposes no assumptions on the server's capabilities, is easier to implement (no concatenation required), and (probably) very effective.

Your approach might be a good idea for sites like YouTube (again assuming the server supports range retrieval requests), where the entire video stream, sometimes multi-gigabyte in length, is delivered in a single container.

**Comment by zmwangx** [comment]
@linhua55 One more thing, as far as I know, while single-threads downloads are easy to resume, resuming segmented downloads is pretty hard. For one thing, you need to keep a lot of (failed or interrupted) [sub]segments around; then, from my naive way of thinking, you have to recalculate ranges for each. Existing tools might be limited. Whether or not `aria2c` supports resuming segmented downloads I don't know, but in the case of `curl`, which I'm quite familiar with, `--continue-at` doesn't play well with `--range`.

**Comment by soimort** [comment]
At this point, the API of `you-get` is kind of broken - there's no single method to extract video URIs, no way to pass in cookies (if needed), or to fetch a specific range of contents, etc. which makes implementing multi-threaded downloading even harder. If you look into `common.py`, there's a lot of mess there (mostly because of the legacy code ported from Python2 that I never maintained seriously)

`you-get` might need a rewrite from scratch to support this. It's on my plan, though.

**Comment by zmwangx** [comment]
> If you look into `common.py`, there's a lot of mess there (mostly because of the legacy code ported from Python2 that I never maintained seriously)

Yeah, that makes sense. I actually looked at `common.py` for ten minutes before reporting this issue, and without digging into details I had a hard time figuring out what each function exactly takes care of — and there seems to be quite a bit of cruft. Docstrings and inline comments are welcome. Also, if there's ever a rewrite, maybe it's worthwhile to use better tooling, e.g., `requests` vs `urllib.request`.

By the way, I believe Python2 support is already broken due to the use of `encoding` keyword argument in `open` ([this line](https://github.com/soimort/you-get/blob/7aeb94fc7d378b48aff7fd14fc428722eda338fe/setup.py#L10)), probably among other things.
